### ISSUE TITLE
[rllib] Checking Serializability of <class 'ray.rllib.agents.trainer_template.A2C'>

### ISSUE THREAD

**Issue by GiovanniGatti** [ray-project/ray#18075]
# What is the problem?

When upgrading from `ray[rllib]==1.5.2` to `ray[rllib]==1.6.0` I'm getting the following error:

```
===========================================================================
Checking Serializability of <class 'ray.rllib.agents.trainer_template.A2C'>
===========================================================================
!!! FAIL serialization: can't pickle function objects
    Serializing '__getstate__' <function build_trainer.<locals>.trainer_cls.__getstate__ at 0x7f2381d71268>...
    !!! FAIL serialization: can't pickle function objects
    Detected 1 global variables. Checking serializability...
        Serializing 'Trainer' <class 'ray.rllib.agents.trainer.Trainer'>...
        !!! FAIL serialization: Could not pickle object as excessively deep recursion required.
            Serializing '__getstate__' <function Trainer.__getstate__ at 0x7f2381ddbe18>...
            !!! FAIL serialization: can't pickle function objects
            Serializing '_allow_unknown_configs' False...
            !!! FAIL serialization: Could not pickle object as excessively deep recursion required.
    Serializing '_allow_unknown_configs' False...
    !!! FAIL serialization: Could not pickle object as excessively deep recursion required.
        Serializing 'denominator' 1...
        !!! FAIL serialization: Could not pickle object as excessively deep recursion required.
            Serializing 'denominator' 1...
            !!! FAIL serialization: Could not pickle object as excessively deep recursion required.
===========================================================================
Variable: 

	FailTuple(_allow_unknown_configs [obj=False, parent=<class 'ray.rllib.agents.trainer.Trainer'>])
FailTuple(denominator [obj=1, parent=1])
FailTuple(__getstate__ [obj=<function Trainer.__getstate__ at 0x7f2381ddbe18>, parent=<class 'ray.rllib.agents.trainer.Trainer'>])

was found to be non-serializable. There may be multiple other undetected variables that were non-serializable. 
Consider either removing the instantiation/imports of these variables or moving the instantiation into the scope of the function/class. 
If you have any suggestions on how to improve this error message, please reach out to the Ray developers on github.com/ray-project/ray/issues/
===========================================================================
Traceback (most recent call last):
  File "src/entry/continuing.py", line 12, in <module>
    from ray.rllib import agents
  File "/<hidden>/python3.7/site-packages/ray/rllib/__init__.py", line 58, in <module>
    _register_all()
  File "/<hidden>/python3.7/site-packages/ray/rllib/__init__.py", line 35, in _register_all
    register_trainable(key, get_trainer_class(key))
  File "/<hidden>/python3.7/site-packages/ray/tune/registry.py", line 73, in register_trainable
    _global_registry.register(TRAINABLE_CLASS, name, trainable)
  File "/<hidden>/python3.7/site-packages/ray/tune/registry.py", line 145, in register
    self._to_flush[(category, key)] = pickle.dumps_debug(value)
  File "/<hidden>/python3.7/site-packages/ray/cloudpickle/__init__.py", line 39, in dumps_debug
    return dumps(obj, *args, **kwargs)
  File "/<hidden>/python3.7/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/<hidden>/python3.7/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 580, in dump
    return Pickler.dump(self, obj)
  File "/<hidden>/python3.7/site-packages/ray/pickle5_files/pickle5/pickle.py", line 485, in dump
    self.save(obj)
  File "/<hidden>/python3.7/site-packages/ray/pickle5_files/pickle5/pickle.py", line 601, in save
    self.save_reduce(obj=obj, *rv)
  File "/<hidden>/python3.7/site-packages/ray/pickle5_files/pickle5/pickle.py", line 689, in save_reduce
    save(func)
  File "/<hidden>/python3.7/site-packages/ray/pickle5_files/pickle5/pickle.py", line 576, in save
    rv = reduce(self.proto)
TypeError: can't pickle function objects
```

Note that it is breaking while loading the RLLib modules `from ray.rllib import agents` (which in my project it is being currently used for type checking).

**OS**
```bash
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.5 LTS
Release:	18.04
Codename:	bionic
```

**Python**
```bash
$ python --version
Python 3.7.1
```
**VirtualEnv config**: [requirements.txt](https://github.com/ray-project/ray/files/7047738/requirements.txt)

# Reproduction
All I did was to change in the requirements file `ray[rllib]==1.5.2` to `ray[rllib]==1.6.0` and install it in a fresh environment. If I rollback this change and repeat the procedure I have no issues.

- [X] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by matthewdeng** [comment]
Hmm, I wasn't able to reproduce with python 3.7.1 myself while running as simple of an install as possible. 
```bash
conda create --name py371  python=3.7.1 
conda activate py371 
pip install "ray[rllib]==1.6.0"
python -c "from ray.rllib import agents"
```
Either way, if a later Python patch version is working for you (3.7.11), are you able to use that instead?

**Comment by GiovanniGatti** [comment]
@matthewdeng, there is always costs and risks changing the Python version. That said, yes, it is possible. I can update it.

However, I'm confident you should be able to reproduce it with the docker image I provided. I cleaned it up a little bit. This is the smallest you need to reproduce it.

```docker
FROM mcr.microsoft.com/azureml/openmpi3.1.2-ubuntu18.04

ENV PYTHON_ENV /.virtualenvs/py37

RUN conda install conda-build \
    && conda create -y -p $PYTHON_ENV python=3.7.1 pip=21.1.2 \
    && conda clean -ya

ENV PATH $PYTHON_ENV/bin:$PATH

RUN python -m pip install --no-cache-dir ray[rllib]==1.6.0
```

**Comment by sikiyu27** [comment]
I have the same problem .Have you solved it?

**Comment by oliver29520** [comment]
I have the same problem. 
python3.6.5
ray1.6.0
ubuntu16.04.6

**Comment by matthewdeng** [comment]
I was able to reproduce this on a clean `Ubuntu 18.0.4` instance and did some further debugging.

More specifically, the error will surface in Python versions that precede `3.6.8` and `3.7.2`. It is interesting to note that these patch versions were released together on 2018-12-24 ([3.6.8](https://www.python.org/dev/peps/pep-0494/#schedule-last-bugfix-release), [3.7.2](https://www.python.org/dev/peps/pep-0537/#id1)) and contain Python bug fixes (I did skim through the patch notes, but nothing stood out as directly related to this issue).

Given this, my recommendation would be to use use Python  `3.6.8+` or `3.7.2+`.
